src/density_interface.jl

```diff
@@ -0,0 +1,19 @@
+@inline DensityInterface.hasdensity(::Distribution) = true
```

Comment:
Last bit, let's check that this should indeed return a Boolean value and not a type, so we are relying on constant propagation instead of Holy-type trait like `Base.IteratorEltype`
```
julia> Base.IteratorEltype([1,1])
Base.HasEltype()
```

Reply:
IMO it should return a `Boolean` and only possibly reevaluated if there are any major issues in some implementation that can't be fixed otherwise. In all cases such as
```julia
f(d) = hasdensity(d) ? g(d) : h(d)
```
the compiler should be able to elide one branch. Generally, I think it is much nicer to work with `Boolean`s since they can be used in if statements etc. whereas otherwise one always has to add a helper function and dispatch.

Reply:
> Last bit, let's check that this should indeed return a Boolean value and not a type

You mean test that returning a Bool is correct (`test_density_interface` checks it)? Or changing DensityInterface itself, and have `hasdensity` return a type?

Hm, constant propagation is supposed to be pretty good nowadays, but ... @devmotion , what do you think?

Reply:
My guess would be that `Val(hasdensity(density_to_test))` should work well nowadays.

Reply:
Ok

Reply:
I trust the compiler here. If we notice any problems, we can reevaluate.

Reply:
Ok, quick test:

```julia
using DensityInterface, Test

struct MyDensity end
@inline DensityInterface.hasdensity(::MyDensity) = true
DensityInterface.logdensityof(::MyDensity, x) = - x^2


get_log_density_value_impl(::Val{true}, d, x) = logdensityof(d, x)
get_log_density_value_impl(::Val{false}, d, x) = "blah, blah"
get_log_density_value(d, x) = get_log_density_value_impl(Val(hasdensity(d)), d, x)

d = MyDensity()
get_log_density_value(d, 4.2)
```

```julia
julia> @inferred get_log_density_value(d, 4.2)
-17.64
```

`@code_warntype get_log_density_value(d, 4.2)` looks fine, too.

Reply:
Also

```julia
julia> hasdensity_as_val(d) = Val(hasdensity(d))
hasdensity_as_val (generic function with 1 method)

julia> @code_llvm hasdensity_as_val(d)
;  @ REPL[28]:1 within `hasdensity_as_val`
define void @julia_hasdensity_as_val_1265() #0 {
top:
  ret void
}
```

Reply:
@mschauer will this work for you?

Reply:
> @mschauer  Ok

Ah, sorry, I overlooked that one. Ok, I'll mark as resolved.